rfcs/0030-ppc64-platforms.md

```diff
@@ -0,0 +1,85 @@
+---
+feature: ppc64-platforms
+start-date: 2018-08-23
+author: CrystalGamma
+co-authors: (find a buddy later to help out with the RFC)
+related-issues: https://github.com/NixOS/nixpkgs/pull/45340 https://github.com/NixOS/nixpkgs/pull/45472 https://github.com/NixOS/nixpkgs/pull/45474 https://github.com/NixOS/nixpkgs/pull/45475
+---
+
+# Summary
+[summary]: #summary
+
+Bring support for ppc64 platforms, initially especially ppc64le, to nixpkgs/NixOS.
+
+# Motivation
+[motivation]: #motivation
+
+NixOS/nixpkgs has long prided itself to be a general-purpose OS/package repository, allowing its packages to be used across a broad range of systems, and providing more packages than ever.
+
+With alternative architectures such as OpenPOWER and RISC-V gaining traction, now is the time to add support for those architectures.
+Specifically motivating this request is the OpenPOWER, or powerpc64le, architecture—and its embodiment in the POWER series of microprocessors from IBM.
+Fully open source commercially available systems have recently become widely available for this platform and there are active Nix users who desire support.
+
+# Detailed design
+[design]: #detailed-design
+
+The first step for implementing support is extending lib/ to recognize powerpc64(le) as a CPU architecture and allowing bootstrap files for the architecture to be (cross-)built and a stdenv based on those bootstrap files to be built.
+This has already been merged in [nixpkgs#45340](https://github.com/NixOS/nixpkgs/pull/45340).
+
+The author of this RFC is already running a NixOS-based system on his POWER9-based computer, so for a basic system without graphical interface, only a handful of packages are known to require changes to build a working system:
+
+* mesa: currently assumes that any non-ARM system is x86. Requires restructuring the driver selection, but to the author's knowledge (since no actual graphical software was tested yet) it should not require any changes to the actual build description. ([nixpkgs#45474](https://github.com/NixOS/nixpkgs/pull/45474))
+* strace: also assumes that any non-ARM system is x86 and so tries to use the m32 personality. Requires checking the available machine personalities for all non-x86 architectures. ([nixpkgs#45472](https://github.com/NixOS/nixpkgs/pull/45472))
+* TeX Live: builds LuaJIT as part of mfluajit, which doesn't work because LuaJIT doesn't support 64-bit Power architectures yet; seems to be possible to disable on that platform without too much impact. ([nixpkgs#45475](https://github.com/NixOS/nixpkgs/pull/45475))
+* OpenBLAS: pulled in by LVM. Needs a target description for ppc64le, but no changes to the build itself. <!-- I ran into a problem with a broken test a while ago, but it seems to be fine right now -->
```

Comment:
OpenBLAS needed by LVM sounds very strange. LLVM maybe? It doesn't seem to be so on x86_64, but I wouldn't be surprised if POWER is different enough.

Reply:
I think it is at least (transitively). It was definitely pulled in by something related to file systems.

Reply:
Strange. It seems it's not required anymore for the system derivation (maybe the dependency was broken somewhere along the line since last month?). Should I remove the bullet point then?

Reply:
Well, if you want scientific calculations (Sage, Julia…) you will need OpenBLAS soon, but you might want to deprioritise it towards the end of the list.

I am not a ppc64 user (neither current nor near future), so I am not in position to give advice on priorities.

Reply:
OK, I guess I will put it after Rust, as something like "Some packages, like OpenBLAS, will need target descriptions for these platforms, but generally don't require changes to the actual build process."